### ISSUE TITLE
Capability to cancel dask jobs started with fire_and_forget

### ISSUE THREAD

**Issue by kyprifog** [dask/distributed#4033]
Ran across a scenario where a job was improperly submitted using "fire_and_forget" to my dask cluster that was too small/wrong level of parallelism and was going to take a long time to finish.    The only way I could find (out of the box) to cancel the job was to blow away the whole cluster and restart it.

It would be nice if Coordination Primitives as described in this (closed) issue:
https://github.com/dask/dask/issues/1183

Could be combined with an approach like described here (explicit stopping condition):
https://stackoverflow.com/questions/49203128/how-do-i-stop-a-running-task-in-dask

To give out of the box ways to have Client apis like:

```
Client.get_jobs
>>  Queue[Future]

Client.cancel_job(future.id)
>>  Cancelled
```

**Comment by mrocklin** [comment]
Hi @kyprifog !

My apologies for the long delay in response.  In principle it looks like you're coming up against two issues:

1.  You need to keep track of the Future's key
2.  It's hard to stop Python threads

As you've pointed out, there are workarounds to each problem, but it's not clear that either workaround should be within Dask itself.  Maybe this belongs in user code.  

> If no one thinks this is a bad idea or if there isn't an easier way to accomplish the same thing I'll get started on this later in the week.

Can you say more about what you are proposing exactly?  A new "Job" abstraction where submitted tasks need to have a while loop that checks status in them?

**Comment by kyprifog** [comment]
More or less I would like to have an optional way to combine the proposed method for arresting a job as proposed in the stack over flow thread with using Coordination Primitives to be able to track track Future's keys so they can be cancelled.

**Comment by kyprifog** [comment]
@mrocklin A (I think pretty common) scenario where this would be useful would be say you submit a job to a cluster that is not set to adaptive scaling, and the job is too big.   If you want to clear that job out of dask to make room for other jobs to run, the only way to do that currently to my knowledge is to kill all the workers and start over.

**Comment by mrocklin** [comment]
I agree with you about the problem, and that it would be good to resolve that problem.  I haven't given this enough thought to know the right way to solve this problem cleanly.  If you have detailed thoughts on how you would resolve this that would be welcome.

**Comment by kyprifog** [comment]
Sure thing, I'll try to sketch something out.

**Comment by bolliger32** [comment]
@kyprifog Curious if you made any progress on this. I have a somewhat related use case that I think would be a useful feature. Occasionally, we map a long-running function across a number of inputs, where the function's main "output" is a side-effect (i.e. a saved file or a saved region of a zarr store). The function then will just return None. We would like to call `fire_and_forget` on these tasks, because we are running them on a google cloud-backed GatewayCluster with preemptible nodes that can disappear. If we keep the futures in scope locally and a worker node disappears (or dies for any other reason) then all of the completed long-running tasks would get run again which we don't want. `fire_and_forget` would get around this issue, but we would also like to know if any of the tasks raise an error. I haven't figured out a way to do this with `fire_and_forget` but our current proposed workaround is the following:

```python
bad_futs = []
exceptions = []
for future, result in dd.as_completed(client.map(function, iterable), with_results=True, raise_errors=False):
    if isinstance(result, tuple) and isinstance(result[0], type) and isinstance(result[1], Exception):
        bad_futs.append(future)
        bad_exceptions.append(result)
```

This should run and then immediately remove any completed task from the scheduler but also hang onto errored tasks. It of course blocks, but I'd imagine someone more skilled than me could make this asynchronous(?)

Anyways, curious if either of you (@mrocklin as well) think this could be a useful feature. I'm thinking an api like:

```python
errored_futures = fire_and_forget(futures, keep_errors=True)
```

Where `errored_futures` is a list that gets appended to by failed futures. Not sure if this would be possible or if this is even common enough to be useful. The workaround is good enough for our use case, but figured I'd bring this up in case it's more broadly useful.
